### ISSUE TITLE
Deprecate asynchronous `SchedulerPlugin.{add|remove}_worker` hooks

### ISSUE THREAD

**Issue by hendrikmakait** [dask/distributed#8487]
Allowing synchronous and asynchronous `SchedulerPlugin.{add|remove}_worker` hooks makes it harder to reason about their execution order (related: #8486). I'm unaware of asynchronous hooks used in practice, so I propose deprecating these. Should there be any legitimate use cases, we can introduce an asynchronous `after_{add|remove}_worker` hook that would allow strict ordering.

**Comment by fjetter** [comment]
The reason why the ordering is hard to reason about is not because they are defined async but because we're not awaiting them individually here https://github.com/dask/distributed/blob/4425516f86d7de9aed06053fc3e21a17fe20efc4/distributed/scheduler.py#L4390-L4399 and https://github.com/dask/distributed/blob/4425516f86d7de9aed06053fc3e21a17fe20efc4/distributed/scheduler.py#L5192-L5213

I don't see a reason to remove async add/remove_worker

**Comment by hendrikmakait** [comment]
> The reason why the ordering is hard to reason about is not because they are defined async but because we're not awaiting them individually here

You're right that we could execute them in order. The one downside of that is that previously synchronous `remove_worker` calls would effectively become asynchronous. For instance, P2P's `remove_worker` hook relied (relies?) on being executed synchronously with the preceding code in `Scheduler.remove_worker` (https://github.com/dask/distributed/issues/7798)

**Comment by hendrikmakait** [comment]
FWIW, we currently make this guarantee about the execution of synchronous `remove_worker` hooks: https://github.com/dask/distributed/blob/774874ebc4b1bea88d0df8288eb0ab67f39004fc/distributed/diagnostics/plugin.py#L190-L193
